### PR TITLE
Update Check Service System

### PR DESCRIPTION
**Pull request by nico-stefani**
|Related issue|
|---|
|#18725|

## Description

This PR adds a new functionality to query the available Wazuh updates from the CTI service. Also, includes a new endpoint for retrieving this information. 

### API Start

```
root@wazuh-master:/var/ossec# tail -f logs/api.log
2023/11/01 19:03:23 INFO: HTTPS is enabled but cannot find the private key and/or certificate. Attempting to generate them
2023/11/01 19:03:24 INFO: Generated private key file in WAZUH_PATH/api/configuration/ssl/server.key
2023/11/01 19:03:24 INFO: Generated certificate file in WAZUH_PATH/api/configuration/ssl/server.crt
2023/11/01 19:03:24 INFO: Checking RBAC database integrity...
2023/11/01 19:03:24 INFO: RBAC database not found. Initializing
2023/11/01 19:03:26 INFO: /var/ossec/api/configuration/security/rbac.db database created successfully
2023/11/01 19:03:26 INFO: RBAC database integrity check finished successfully
2023/11/01 19:03:33 INFO: Listening on 0.0.0.0:55000..
2023/11/01 19:03:33 INFO: Populating installation UID...
2023/11/01 19:03:33 INFO: Getting updates information...
```

### Request examples

* Basic request
```
curl -k --silent -H  "Authorization: Bearer $TOKEN" https://localhost:55050/manager/version/check | jq
{
  "data": {
    "last_check_date": "2023-11-01T19:15:08.630419+00:00",
    "current_version": "v4.8.0",
    "update_check": true,
    "last_available_major": {
      "tag": "v5.0.0",
      "description": null,
      "title": "Wazuh 5.0.0",
      "published_date": "2023-10-05T12:48:00Z",
      "semver": {
        "major": 5,
        "minor": 0,
        "patch": 0
      }
    },
    "last_available_minor": {
      "tag": "v4.9.1",
      "description": null,
      "title": "Wazuh 4.9.1",
      "published_date": "2023-10-05T12:47:00Z",
      "semver": {
        "major": 4,
        "minor": 9,
        "patch": 1
      }
    },
    "last_available_patch": {
      "tag": "v4.8.2",
      "description": null,
      "title": "Wazuh 4.8.2",
      "published_date": "2023-10-05T12:46:00Z",
      "semver": {
        "major": 4,
        "minor": 8,
        "patch": 2
      }
    }
  },
  "error": 0
}
```

* Request with `force_query=true`

```
curl -k --silent -H  "Authorization: Bearer $TOKEN" https://localhost:55050/manager/version/check\?force_query\=true | jq
{
  "data": {
    "last_check_date": "2023-11-01T19:17:35.054740+00:00",
    "current_version": "v4.8.0",
    "update_check": true,
    "last_available_major": {
      "tag": "v5.0.0",
      "description": null,
      "title": "Wazuh 5.0.0",
      "published_date": "2023-10-05T12:48:00Z",
      "semver": {
        "major": 5,
        "minor": 0,
        "patch": 0
      }
    },
    "last_available_minor": {
      "tag": "v4.9.1",
      "description": null,
      "title": "Wazuh 4.9.1",
      "published_date": "2023-10-05T12:47:00Z",
      "semver": {
        "major": 4,
        "minor": 9,
        "patch": 1
      }
    },
    "last_available_patch": {
      "tag": "v4.8.2",
      "description": null,
      "title": "Wazuh 4.8.2",
      "published_date": "2023-10-05T12:46:00Z",
      "semver": {
        "major": 4,
        "minor": 8,
        "patch": 2
      }
    }
  },
  "error": 0
}
```

* Request when service is disabled

```
curl -k --silent -H  "Authorization: Bearer $TOKEN" https://localhost:55050/manager/version/check | jq
{
  "data": {
    "last_check_date": "",
    "current_version": "v4.8.0",
    "update_check": false,
    "last_available_major": {},
    "last_available_minor": {},
    "last_available_patch": {}
  },
  "error": 0
}
```

* Request when an error communicating with CTI

```
curl -k --silent -H  "Authorization: Bearer $TOKEN" https://localhost:55050/manager/version/check | jq
{
  "title": "Wazuh Internal Error",
  "detail": "Error in CTI service request: Cannot connect to host cti.wazuh.com:443 ssl:default [Name or service not known]",
  "dapi_errors": {
    "master-node": {
      "error": "Error in CTI service request: Cannot connect to host cti.wazuh.com:443 ssl:default [Name or service not known]",
      "logfile": "WAZUH_HOME/logs/api.log"
    }
  },
  "error": 2100
}
```

* Request on a worker node

```
curl -k --silent -H  "Authorization: Bearer $TOKEN_WORKER" https://localhost:55051/manager/version/check | jq
{
  "title": "Resource Not Found",
  "detail": "API Endpoint only available on master node",
  "error": 902
}
```

### CODE REVIEW

**Comment by nico-stefani** [comment]
I've created [feature/18725-update-check-service-system-bis](https://github.com/wazuh/wazuh/tree/feature/18725-update-check-service-system-bis) and cherry-picked eed98951c8f3459950c9ce8325b192b89d683a36 to execute the *Integration tests for API - Tier 0 and 1 / build (pull_request)* pipelines. 
[Here](https://github.com/wazuh/wazuh/actions/runs/6724448254/job/18276737763) are the results :green_circle:

File: api/api/signals.py [hunk 1]
```diff
@@ -1,8 +1,138 @@
...
+
+def _get_current_version() -> str:
+    """Return the version of running Wazuh instance.
+
+    Returns
+    -------
+    str
+        Wazuh version in X.Y.Z format.
+    """
+    return wazuh.__version__
```
Comment (Selutario) [inline api/api/signals.py:72]: These are no longer being used

File: api/api/signals.py [hunk 2]
```diff
@@ -1,8 +1,138 @@
 # Copyright (C) 2015, Wazuh Inc.
 # Created by Wazuh, Inc. <info@wazuh.com>.
 # This program is a free software; you can redistribute it and/or modify it under the terms of GPLv2
+import asyncio
+import logging
+import os
+import ssl
```
Comment (Selutario) [inline api/api/signals.py:7]:
```suggestion
```

File: api/api/signals.py [hunk 3]
```diff
@@ -1,8 +1,138 @@
...
+import uuid
+from functools import wraps
+from typing import AsyncGenerator, Callable
+
+import aiohttp
+import certifi
```
Comment (Selutario) [inline api/api/signals.py:13]:
```suggestion
```

File: api/api/signals.py [hunk 4]
```diff
@@ -1,8 +1,138 @@
...
+from aiohttp import web
+
+import wazuh
```
Comment (Selutario) [inline api/api/signals.py:16]:
```suggestion
```

File: api/test/integration/test_manager_endpoints.tavern.yaml [hunk 5]
```diff
@@ -1560,6 +1560,271 @@ stages:
...
+        error: 0
+        data:
+          affected_items:
+            - !anystr
+          failed_items: []
+          total_affected_items: 1
+          total_failed_items: 0
+    delay_after: 50
+
+  - name: Get available updates
```
Comment (Selutario) [inline api/test/integration/test_manager_endpoints.tavern.yaml:1802]:
```suggestion
  - name: Try to get available updates
```

File: api/test/integration/test_manager_endpoints.tavern.yaml [hunk 6]
```diff
@@ -1560,6 +1560,271 @@ stages:
...
+      url: "{protocol:s}://{host:s}:{port:d}/manager/version/check"
+      method: GET
+      headers:
+        Authorization: "Bearer {test_login_token}"
+    response:
+      status_code: 500
+      json:
+        error: 2100
+
+  - name: Get available updates with force option
```
Comment (Selutario) [inline api/test/integration/test_manager_endpoints.tavern.yaml:1814]:
```suggestion
  - name: Try to get available updates with force option
```

**Review by Selutario — CHANGES_REQUESTED** [review]

**Review by Selutario — APPROVED** [review]
LGTM. 

Failing checks are:
- **[Framework](https://github.com/wazuh/wazuh/pull/19952/checks?check_run_id=18494734783) and [API](https://github.com/wazuh/wazuh/pull/19952/checks?check_run_id=18518909484) unit tests:** Probably failed due to [changes in Jenkins](https://github.com/wazuh/wazuh-jenkins/pull/5268) in order to use Python 3.10. Should be fixed after rebasing with 4.8.0. The same tests were also run in Github Actions, where they passed successfully:
  - [Framework tests](https://github.com/wazuh/wazuh/actions/runs/6801939061)
  - [API tests](https://github.com/wazuh/wazuh/actions/runs/6801939083)
- **[Integration tests for API - Tier 0 and 1](https://github.com/wazuh/wazuh/actions/runs/6801939072):** The same tests passed successfully [here](https://github.com/wazuh/wazuh/pull/19952#issuecomment-1789633972) a week ago. The error does not seem related to the changes in this PR. 
  ```
  ValueError: Error when executing restart in daemon None. Exit status: 1
  ```
  In any case, please @nico-stefani open an issue to investigate it separately. 
- **[Build Linux x86_64/amd64 agent/manager](https://github.com/wazuh/wazuh/pull/19952/checks?check_run_id=18494182556):** The failures in those checks have nothing to do with the changes in this PR.
- **[Integration API endpoints test](https://github.com/wazuh/wazuh/pull/19952/checks?check_run_id=18495558248):** 
  - **test_agent_GET_endpoints:** Failed tests are related to [#20002](https://github.com/wazuh/wazuh/issues/20002), and maybe to [#20143](https://github.com/wazuh/wazuh/issues/20143). No changes were made in agent API endpoints.
  - **test_cluster_endpoints:** There were no changes in cluster API endpoints. The failed tests are related to [#20002](https://github.com/wazuh/wazuh/issues/20002) and [#18284](https://github.com/wazuh/wazuh/issues/18284).
  - **test_manager_endpoints:** There are failed tests after restarting Wazuh as a step of one of the Update check system tests. Restarting is taking more than 50s to complete, making it fail. Related to [#20143](https://github.com/wazuh/wazuh/issues/20143).

**Pull request merged**